File: csharp/11_Pull/Months.cs
using System;
using System.Collections.Generic;
using Common;
using Pull;

namespace Pull.Months
{
    public class Months
    {
        private readonly IList<ValuesOfMonth> months = new List<ValuesOfMonth>();
        private readonly IDictionary<Month, ValuesOfMonth> monthsInMap = new Dictionary<Month, ValuesOfMonth>();

        public Months(IList<BalancesOfMonth> balancesOfOneAccount, IList<Transaction> transactions)
        {
            Init(balancesOfOneAccount, transactions);
        }

        public ValuesOfMonth ForDate(DateTime date)
        {
            return monthsInMap[new Month(date)];
        }

        private void Init(IList<BalancesOfMonth> balancesOfOneAccount, IList<Transaction> transactions)
        {
            ValuesOfMonth month = null;
            foreach (BalancesOfMonth balancesOfMonth in balancesOfOneAccount)
            {
                month = new ValuesOfMonthWithCaching(balancesOfMonth.Date, month != null ? (IValuesOfMonth)month : new DummyValuesOfMonth());
                months.Add(month);
                monthsInMap[month.Month] = month;
            }

            AllocateTransactionsToMonths(transactions);
        }

        private void AllocateTransactionsToMonths(IList<Transaction> transactions)
        {
            foreach (Transaction transaction in transactions)
            {
                ForDate(transaction.Date).AddTransaction(transaction);
            }
        }
    }
}

File: java/11Pull/src/pull/months/Months.java
package pull.months;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.joda.time.LocalDate;
import org.joda.time.YearMonth;

import common.BalancesOfMonth;
import common.Transaction;

public class Months
{

	private final List<ValuesOfMonth> months = new ArrayList<ValuesOfMonth>();
	private final Map<YearMonth, ValuesOfMonth> monthsInMap = new HashMap<YearMonth, ValuesOfMonth>();

	public Months(List<BalancesOfMonth> balancesOfOneAccount, List<Transaction> transactions)
	{
		super();
		init(balancesOfOneAccount, transactions);
	}

	private void init(List<BalancesOfMonth> balancesOfOneAccount, List<Transaction> transactions)
	{
		ValuesOfMonth month = null;
		for (BalancesOfMonth balancesOfMonth : balancesOfOneAccount)
		{
			month = new ValuesOfMonthWithCaching(balancesOfMonth.getDate(), month != null ? month : new InitialValuesOfMonth());
			getMonths().add(month);
			monthsInMap.put(month.getYearMonth(), month);
		}

		allocateTransactionsToMonths(transactions);
	}

	public List<ValuesOfMonth> getMonths()
	{
		return months;
	}

	public ValuesOfMonth forDate(LocalDate date)
	{
		return monthsInMap.get(new YearMonth(date));
	}

	private void allocateTransactionsToMonths(List<Transaction> transactions)
	{
		for (Transaction transaction : transactions)
		{
			forDate(transaction.getDate()).addTransaction(transaction);
		}
	}

}
